### ISSUE TITLE
Use std::unique_ptr to express ownership

### ISSUE THREAD

**Issue by drwells** [IBAMR/IBAMR#579]
From #570: we should use `std::unique_ptr` when moving pointer ownership around in `HierarchyIntegrator`.

I think its okay if we break this interface for the sake of progress.

**Comment by drwells** [comment]
I think we can preserve compatibility and make the semantics clearer by overloading the remaining functions, i.e., have both
```cpp
void registerGhostfillRefineAlgorithm(const std::string& ,
                                      SAMRAI::tbox::Pointer<SAMRAI::xfer::RefineAlgorithm<NDIM> > ,
                                      std::unique_ptr<SAMRAI::xfer::RefinePatchStrategy<NDIM> );

void registerGhostfillRefineAlgorithm(const std::string& ,
                                      SAMRAI::tbox::Pointer<SAMRAI::xfer::RefineAlgorithm<NDIM> > );
```
since we always assume ownership of the `RefinePatchStrategy` object anyway.

**Comment by drwells** [comment]
@marshallrdavey Are you willing to take a look at this?

**Comment by marshallrdavey** [comment]
@drwells I'm running into issues in IBHierachyInegrator and others that pass a pointer into registerGhostfillRefineAlgorithm instead of unique_ptr. What would you like to do in those cases?

**Comment by drwells** [comment]
If the class inherits from `HierarchyIntegrator` then the parent class is responsible for calling the destructor. IBHierarchyIntegrator stores a pointer to the object itself.

We can probably do the following:
```cpp
std::unique_ptr<RefinePatchStrategy<NDIM> > rps(d_u_phys_bdry_op);
registerGhostfillRefineAlgorithm(d_object_name + "::u", d_u_ghostfill_alg, std::move(rps));
```
we should also update the documentation to clarify that `d_u_phys_bdry_op` is managed by the parent class since it is registered as a ghostfill refine algorithm.

**Comment by drwells** [comment]
This works only because we create the object (via `new`) in the inheriting class but delete it (by putting it in a `unique_ptr`) in the base class.

**Comment by drwells** [comment]
Fixed by #849.
